src/rgw/rgw_op.cc

```diff
@@ -4180,9 +4181,41 @@ void RGWCompleteMultipart::execute()
   meta_obj.set_in_extra_data(true);
   meta_obj.index_hash_source = s->object.name;
 
+  auto raw_oid = meta_obj.bucket.bucket_id + "_" + meta_obj.get_object();
+
+  /*take a cls lock on meta_obj to prevent racing completions (or retries)
+    from deleting the parts*/
+  librados::ObjectWriteOperation op;
+  librados::IoCtx ioctx;
+  rados::cls::lock::Lock l("RGWCompleteMultipart");
+  int max_lock_secs_mp = s->cct->_conf->rgw_mp_lock_max_time;
+  utime_t time(max_lock_secs_mp, 0);
+
+  op.assert_exists();
+  l.set_duration(time);
+  l.lock_exclusive(&op);
+
+  op_ret = store->get_obj_ioctx(meta_obj, &ioctx);
```

Comment:
iiuc, you want to move the lock structure setup after get_obj_ioctx?  that's fine, but I don't think there's a critical section here